neps/nep-0455.md

```diff
@@ -0,0 +1,191 @@
+
+If we don't address the undercharging problems, we increase the risks that they will be exploited and also make it harder to deploy Flat Storage safely.
+
+## Alternatives
+
+### Increase the gas costs for undercharged operations
+We could increase the gas costs for the operations that are undercharged to match the computational time it takes to process them according to the rule 1ms = 1TGas.
+
+Pros:
+- Does not require any new code or design work (but still requires a protocol version bump)
+- Security implications are well-understood
+
+Cons:
+- Can break contracts that rely on current gas costs, in particular steeply increasing operating costs for the most active users of the blockchain (aurora and sweat)
+- Doing this safely and responsibly requires prior consent by the affected parties which is hard to do without disclosing the security-sensitive information about undercharging in public
+
+In case of flat storage specifically, using this approach will result in a large increase in storage write costs (breaking many contracts) to enable safe deployment of read-only flat storage and later a correction of storage write costs when flat storage for writes is rolled out.
+With compute costs, we will be able to roll out the read-only flat storage with minimal impact on deployed contracts.
+
+### Adjust the gas chunk limit
+We could continuously measure the chunk production time in nearcore clients and compare it to the gas burnt.
+If the chunk producer observes undercharging, it decreases the limit.
+If there is overcharging, the limit can be increased up to a limit of at most 1000 Tgas.
+
+Pros:
+- Prevents moderate undercharging from stalling the network
+- No protocol change necessary (as this feature is already [a part of the protocol](https://nomicon.io/Economics/Economic#transaction-fees)), we could easily experiment and revert if it does not work well
+
+Cons:
+- Very broad granularity --- undercharging in one parameter affects all users, even those that never use the undercharged parts
+- Dependence on validator hardware --- someone running overspecced hardware will continuously want to increase the limit, others might run with underspecced hardware and continuously want to decrease the limit
+- Malicious undercharging attacks are unlikely to be prevented by this --- a single 10x undercharged receipt still needs to be processed using the old limit.
+Adjusting 0.1% per block means 100 chunks can only change by a maximum of 1.1x and 1000 chunks could change up to x2.7
+- Conflicts with transaction and receipt limit --- A transaction or receipt can (today) use up to 300Tgas.
+The effective limit per chunk is `gas_limit` + 300Tgas since receipts are added to a chunk until one exceeds the limit and the last receipt is not removed.
+Thus a gas limit of 0gas only reduces the effective limit from 1300Tgas to 300Tgas, which means undercharging of > 4.33x is not covered.
```

Comment:
can you briefly explain how 4.33x is computed?

Reply:
It rewrote this in a way that's easier to understand, PTAL

Reply:
much better!